src/fundus/scraping/pipeline.py

```diff
@@ -62,6 +61,102 @@ def __init__(self, *scrapers: Scraper):
+        """Async variant of the crawl() method.
+
+        See docstring for crawl(). for detailed information about the parameters.
+
+        Args:
+            max_articles (Optional[int]): Number of articles to crawl. Defaults to None.
+            error_handling (Literal["suppress", "catch", "raise"]): Set error handling. Defaults to "suppress".
+            only_complete (Union[bool, ExtractionFilter]): Set extraction filters. Defaults to True
+            delay (Optional[Union[float, Delay]]): Set delay time between article batches. Defaults to None.
+            url_filter (Optional[URLFilter]): Set URLFilter. Defaults to None.
+            only_unique (bool): If true return only unique responses. Defaults to True.
+
+        Returns:
+            AsyncIterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        def build_extraction_filter() -> Optional[ExtractionFilter]:
+            if isinstance(only_complete, bool):
+                return (
+                    None
+                    if only_complete is False
+                    else lambda extracted: not all(
+                        bool(v) if not isinstance(v, Exception) else False for k, v in extracted.items()
+                    )
+                )
+            else:
+                return only_complete
+
+        def build_delay() -> Optional[Delay]:
+            if isinstance(delay, float):
+
+                def constant_delay() -> float:
+                    return delay  # type: ignore[return-value]
+
+                return constant_delay
+            else:
+                return delay
+
+        def build_unique_url_filter() -> URLFilter:
+            return lambda url: url in response_cache
+
+        # build filters and delay. this is for readability and typeguard reasons
+        extraction_filter = build_extraction_filter()
+        unique_url_filter = build_unique_url_filter() if only_unique else None
+        final_delay = build_delay()
+
+        response_cache: Set[str] = set()
+
+        for scraper in self.scrapers:
+            for source in scraper.sources:
+                if url_filter:
+                    source.add_url_filter(url_filter=url_filter)
+                if unique_url_filter:
+                    source.add_url_filter(url_filter=unique_url_filter)
+
+        async_article_iterators: List[AsyncIterator[Optional[Article]]] = [
+            scraper.scrape(
+                error_handling=error_handling,
+                extraction_filter=extraction_filter,
+            )
+            for scraper in self.scrapers
+        ]
+
+        # we use this custom variant of interleave_longest in order to be able
+        # to delay the program flow between batches
+        async def _async_article_interleave_longest() -> AsyncIterator[Article]:
+            batches: AsyncIterator[Tuple[Optional[Article], ...]] = async_zip_longest(*async_article_iterators)
+            start_time = time.time()
```

Comment:
My point is more about avoiding accidental redefinitions of variables, since then mypy will catch that a variable has been redefined. But it's just a minor nitpick.

Reply:
Mypy would catch the redefinition anyways. You do not need a type hint for mypy to recognize this.

``` python
local_time = time.time()
local_time = "loacl_time"
```

```
error: Incompatible types in assignment (expression has type "str", variable has type "float")  [assignment]
```

Reply:
Yes, but not if there are the same type.

Reply:
Okay, thin I think I don't get what you mean. Could you give me an example?

Reply:
Sure consider this example:

```python
time: int = 0

# do stuff using "time" in a long function

# do more stuff using a different "time" variable but accidentally use the same name

time: int = 123

# do more stuff

# now we need the original time here again but missed that we re-declared it and just use "time" as is
# this would now be a logical error
```

On this code, mypy would throw the following error

```console
main.py:7: error: Name "time" already defined on line 1  [no-redef]
Found 1 error in 1 file (checked 1 source file)
```

So with this, we would catch accidental re-definitions. Again this is really nitpicky and should not stop this PR from getting merged. Just a minor note.